### ISSUE TITLE
Support getting default defines and include paths from the used compiler

### ISSUE THREAD

**Issue by mhoeher** [niosus/EasyClangComplete#370]
Currently, when using ECC with CMake or a `compile_commands.json`, the used flags that are passed to the compiler include only the *explicitly set ones that are passed via the command line*. However, usually compilers also have some built-in include paths and defines that are set regardless of what flags are passed to the compiler by the user.

I guess when developing "host" applications, this usually is not a big problem, because the *usual* defines one would use are mostly the same across different compilers. However, especially when working with toolchains for embedded platforms, the target compiler often sets some special defines which then leads to ECC not being able to provide accurate completions in some cases.

Long story short:
It is possible to retrieve the defines that are set by default like this (also works for LLVM based compilers):

```bash
g++ -dM -E -x c++ -std=c++98 /dev/null
```

And the default include paths can be found by invoking a compiler like this (again this also works with LLVM based compilers):

```bash
g++ -v -x c++ -E /dev/null
```

It would be cool if ECC could query these defaults and append them explicitly to the flags passed to clang to get the auto-completions (of course, that should be optional so in case a compiler is used which does not allow this kind of retrieval or where the returned flags conflict with the ones clang uses internally, ECC remains functional).

As last time, I would volunteer to provide an implementation of this if this kind of feature is considered okay ;-)

**Comment by niosus** [comment]
Hmm, this is a new angle for me. I fell like this is actually a pretty neat idea to have that.
All the hassle with the initial values for the `common_flags` setting is there just because I did not think about this. 

I feel like it would be great to use the `clang_binary` from the settings and parse the output of the command you posted to get the initial value for the `common_flags`. 

Great idea! If you want to implement it, I will be more than happy to review a PR on this.

**Comment by mhoeher** [comment]
Okay, I'll try to implement something the following days ;-)

**Comment by niosus** [comment]
Awesome! :) Feel free to follow up in case something in the code is unclear or if you feel like some design decision gets in your way.

**Comment by niosus** [comment]
Also, please reference #196 as it is related.

**Comment by kblomqvist** [comment]
I have been working on this. My goal has been a PR to support SCons.
https://gist.github.com/kblomqvist/d5ebe28d2604420d5d08553b4c575dbf (Py3 only)

**Comment by mhoeher** [comment]
@kblomqvist thanks for the hint. I might borrow some of your code ;-)

**Comment by kblomqvist** [comment]
I think I manage to dump build info from SConstruct file used by SCons build system (gist updated). I think we can reuse GCC related functions to get system info. I'm waiting @niosus to sync on this. Will then update my code and make a PR to support SCons.

**Comment by niosus** [comment]
@kblomqvist let's move the discussion on SCons build system to a new issue if you would want to add this as a PR. We can discuss this further there if that is ok with you.

**Comment by mhoeher** [comment]
Hi @niosus,

I just created a merge requests with the preliminary changes. For now, the built in flags are used when using a `compile_commands.json` file or when using `cmake` directly. For other methods, the built-ins are not yet used.

I tried the changes at work with our embedded project and there it's working fine so far. `clang` now correctly sees the embedded compiler's specific defines and hence code completion as well as code analysis on save are much better (before, for some functions the "embedded" view used to see some functions targeted for the host which caused some false positives in error reporting).

Comments for improvements are welcome, dunno if the solution as it is like now is okay, so let me know what you think.
